prerendering-state.md

```diff
@@ -0,0 +1,140 @@
+# Prerendering State API
+
+## Summary
+
+We propose introducing a boolean property `document.prerendering` and associated change event to distinguish
+prerendering browsing contexts from regular ones.
+
+_Note: This is based on the more general discussion in this [background
+doc](https://docs.google.com/document/d/1Xzw0k8DgltI2ohapuDKmjRZLv7NVrRFGusW8IBtiCT0/edit?usp=sharing)._
+
+## Background
+
+Pages need to know when they're being rendered inside a [prerendering browsing context](browsing-context.md).
+Some examples of divergent behavior:
+
+* Avoid fetching large video resources
+* Avoid measuring ad impressions
+* Avoid running JavaScript animations
+* Avoid measuring user interactions for analytics
+* Avoid starting server-coordinated behavior (e.g. ticket purchase countdown).
+
+A previous iteration of prerendering used [page-visibility](https://w3c.github.io/page-visibility/) to denote a
+prerendering context. However, this was unimplemented and [removed from the
+specification](https://github.com/w3c/page-visibility/issues/42).
+
+With the re-introduction of prererendering, we need to bring back some form of this API.
+
+## Why not `visibilityState=='prerender'`?
+
+The most straightforward way to do this would be to bring back the previously specified `document.visibilityState ==
+'prerender'`.  Doing so has a number of drawbacks:
+
+### Not mutually-exclusive
+
+In actuality, some of the above use cases are interested in visibility rather than prerendering state.  For example, an
+ordinary (i.e. non-prerendering) page that's hidden would likely want to "avoid running JavaScript animations" and
+"avoid fetching large video resources". Thus, pages would have to consider both the `hidden` state as well as a
+re-introduced `prerender` to get the desired behavior.
+
+However, the [portals proposal](https://github.com/WICG/portals/) means that `prerender` isn't mutually exclusive
+with `visible` either. In that case, we do want to "run JavaScript animations" and "fetch large resources" since it'll
+be visible to the user, despite being in a prerendering browsing context (and all the restrictions that come with that).
+
+### Web-compatibility
+
+Related to the above point, a common pattern on the web is to assume only two visibility states. For example, code such
```

Comment:
Hm, this is rather unfortunate since `prerender` was only removed from the spec in 2019 and WebKit just removed it from trunk a few weeks ago in https://bugs.webkit.org/show_bug.cgi?id=215851. It's unfortunate that visibilityState is not a boolean, but we don't want to add new types because sites are treating it as one. But the argument above about prerendering not being mutually exclusive with visibility and hidden makes sense.